File: custom-packages/cordova-plugins/simplepos-plugin-posprinter/src/android/posprinter.java
/**
 */
package com.cordova.plugin;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import android.util.Log;

import com.epson.epos2.Epos2Exception;
import com.epson.epos2.discovery.DeviceInfo;
import com.epson.epos2.discovery.Discovery;
import com.epson.epos2.discovery.DiscoveryListener;
import com.epson.epos2.discovery.FilterOption;
import com.epson.epos2.printer.Printer;
import com.epson.epos2.printer.PrinterStatusInfo;
import com.epson.epos2.printer.ReceiveListener;

public class posprinter extends CordovaPlugin {
    private static final String TAG = "posprinter";
    private CallbackContext callbackContext = null;
    private Printer printer = null;

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }

    public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        this.callbackContext = callbackContext;

        if (action.equals("startDiscover")) startDiscovery(callbackContext);
        else if (action.equals("stopDiscover")) stopDiscovery(callbackContext);
        else if (action.equals("connectPrinter")) connectPrinter(args, callbackContext);
        else if (action.equals("disconnectPrinter")) disconnectPrinter();
        else if (action.equals("addText")) addText(args, callbackContext);
        else if (action.equals("addTextAlign")) addTextAlign(args, callbackContext);
        else if (action.equals("addTextSize")) addTextSize(args, callbackContext);
        else if (action.equals("addFeedLine")) addFeedLine(args, callbackContext);
        else if (action.equals("addBarcode")) addBarcode(args, callbackContext);
        else if (action.equals("addCut")) addCut(args, callbackContext);
        else if (action.equals("addTextStyle")) addTextStyle(args, callbackContext);
        else if (action.equals("addPulse")) addPulse(args, callbackContext);
        else if (action.equals("print")) print(callbackContext);

      return true;
    }

    private void startDiscovery(final CallbackContext callbackContext) {
        FilterOption mFilterOption = new FilterOption();
        mFilterOption.setDeviceType(Discovery.TYPE_PRINTER);
        mFilterOption.setEpsonFilter(Discovery.FILTER_NAME);
        try {
            Discovery.start(webView.getContext(), mFilterOption, discoveryListener);
        } catch (Epos2Exception e) {
            Log.e(TAG, "Error discovering printer: " + e.getErrorStatus(), e);
            callbackContext.error("Error discovering printer: " + e.getErrorStatus());
        }
    }

    private void stopDiscovery(final CallbackContext callbackContext) {
        while (true) {
            try {
                Discovery.stop();
                PluginResult result = new PluginResult(Status.OK, true);
                callbackContext.sendPluginResult(result);
                break;
            }
            catch (Epos2Exception e) {
                if (e.getErrorStatus() != Epos2Exception.ERR_PROCESSING) {
                    PluginResult result = new PluginResult(Status.ERROR, false);
                    callbackContext.sendPluginResult(result);
                    return;
                }
            }
        }
    }

    private void connectPrinter(final JSONArray args, final CallbackContext callbackContext) {
        Log.d(TAG, "Connecting Printer");


        try {
          int type = args.getInt(0);

          printer = new Printer(type, Printer.MODEL_ANK, webView.getContext());
            printer.setReceiveEventListener(receiveListener);
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error creating printer: " + e.getErrorStatus());
            Log.e(TAG, "Error creating printer: " + e.getErrorStatus(), e);
            return;
        } catch (JSONException e) {
          callbackContext.error("Error getting target: " + e.getCause());
          Log.e(TAG, "Error connecting printer", e);
          return;
        }

      try {
          String target = args.getString(1);

          printer.connect("TCP:" + target , Printer.PARAM_DEFAULT);
        } catch (Epos2Exception e) {
            callbackContext.error("Error connecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error connecting printer: " + e.getErrorStatus(), e);
            return;
        } catch (JSONException e) {
            callbackContext.error("Error getting target: " + e.getCause());
            Log.e(TAG, "Error connecting printer", e);
            return;
        }

        try {
            printer.beginTransaction();
        } catch (Epos2Exception e) {
            callbackContext.error("Error beginning transaction");
            Log.e(TAG, "Error beginning transaction", e);
            return;
        }

        PluginResult result = new PluginResult(Status.OK, "Done connecting");
        callbackContext.sendPluginResult(result);
    }

    private void disconnectPrinter() {
        if (printer == null) {
            return;
        }

        try {
            printer.endTransaction();
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error ending transaction: " + e.getErrorStatus());
            Log.e(TAG, "Error ending transaction: " + e.getErrorStatus(), e);
            return;
        }

        try {
            printer.disconnect();
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error disconnecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error disconnecting printer: " + e.getErrorStatus(), e);
            return;
        }

        printer.clearCommandBuffer();
        printer.setReceiveEventListener(null);
        printer = null;

        PluginResult result = new PluginResult(Status.OK, true);
        callbackContext.sendPluginResult(result);
    }

  private void addText(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      String data = array.getString(0);
      printer.addText(data);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addTextAlign(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int data = array.getInt(0);

      printer.addTextAlign(data);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addTextSize(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int width = array.getInt(0);
      int height = array.getInt(1);

      printer.addTextSize(width, height);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addFeedLine(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int line = array.getInt(0);

      printer.addFeedLine(line);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addBarcode(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      String data = array.getString(0);
      int type = array.getInt(1);
      int hri = array.getInt(2);
      int font = array.getInt(3);
      int width = array.getInt(4);
      int height = array.getInt(5);

      printer.addBarcode(data, type, hri, font, width, height);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addCut(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int type = array.getInt(0);

      printer.addCut(1);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addTextStyle(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int reverse = array.getInt(0);
      int ul = array.getInt(1);
      int em = array.getInt(2);
      int color = array.getInt(3);

      printer.addTextStyle(reverse, ul, em, color);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

  private void addPulse(final JSONArray array, final CallbackContext callbackContext) {
    if (printer == null) {
      callbackContext.error("printer not found");
      return;
    }

    try {
      int drawer = array.getInt(0);
      int time = array.getInt(1);

      printer.addPulse(drawer, time);

      callbackContext.success();

    } catch (JSONException e) {
      callbackContext.error("Error getting data");
      Log.e(TAG, "Error getting data", e);
    } catch (Epos2Exception e) {
      callbackContext.error("Error printing");
      Log.e(TAG, "Error printing", e);
      try {
        printer.disconnect();
      } catch (Epos2Exception ex) {
        callbackContext.error("Error disconnecting");
        Log.e(TAG, "Error disconnecting", ex);
      }
    }
  }

    private void print(final CallbackContext callbackContext) {
      if (printer == null) {
        callbackContext.error("printer not found");
        return;
      }

      try {

        PrinterStatusInfo status = printer.getStatus();

        if (!isPrintable(status)) {
          callbackContext.error("Error printing: printer is not printable");
          Log.e(TAG, "Error printing: printer is not printable");

          try {
            printer.disconnect();
          } catch (Epos2Exception ex) {
            callbackContext.error("Error disconnecting");
            Log.e(TAG, "Error disconnecting", ex);
          }
          return;
        }

        printer.sendData(Printer.PARAM_DEFAULT);
      } catch (Epos2Exception e) {
        callbackContext.error("Error printing");
        Log.e(TAG, "Error printing", e);
        try {
          printer.disconnect();
        } catch (Epos2Exception ex) {
          callbackContext.error("Error disconnecting");
          Log.e(TAG, "Error disconnecting", ex);
        }
      }
    }

    private DiscoveryListener discoveryListener = new DiscoveryListener() {
        @Override
        public void onDiscovery(final DeviceInfo deviceInfo) {
            JSONObject item = new JSONObject();

            try {
                item.put("deviceName", deviceInfo.getDeviceName());
                item.put("target", deviceInfo.getTarget());
                item.put("ipAddress", deviceInfo.getIpAddress());
                item.put("macAddress", deviceInfo.getMacAddress());
                item.put("deviceType", deviceInfo.getDeviceType());
                item.put("bdAddress", deviceInfo.getBdAddress());
            } catch (JSONException e) {
                callbackContext.error("Error building device info");
            }

            callbackContext.success(item);
        }
    };

    private ReceiveListener receiveListener = new ReceiveListener() {
        @Override
        public void onPtrReceive(final Printer printer, final int code, final PrinterStatusInfo status, final String printJobId) {
            Log.d(TAG, "print success. status: " + status.getErrorStatus());

            new Thread(new Runnable() {
                @Override
                public void run() {
                    disconnectPrinter();
                }
            }).start();
        }
    };

    private boolean isPrintable(PrinterStatusInfo status) {
        if (status == null) {
            return false;
        }

        if (status.getConnection() == Printer.FALSE) {
            return false;
        }
        else if (status.getOnline() == Printer.FALSE) {
            return false;
        }

        return true;
    }
}

File: custom-packages/cordova-plugins/simplepos-plugin-posprinter/src/ios/posprinter.m
#import "posprinter.h"

#import <Cordova/CDVAvailability.h>

@interface posprinter()<Epos2DiscoveryDelegate, Epos2PtrReceiveDelegate>
@end

@implementation posprinter

- (void)pluginInitialize {
    printerSeries = EPOS2_TM_T82;
    lang = EPOS2_MODEL_ANK;
}

- (void)startDiscover:(CDVInvokedUrlCommand *)command {
    self.discoverCallbackId = command.callbackId;

    Epos2FilterOption *filteroption_ = [[Epos2FilterOption alloc] init];
    [filteroption_ setDeviceType:EPOS2_TYPE_PRINTER];

    int result = [Epos2Discovery start:filteroption_ delegate:self];
    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in startDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

- (void) onDiscovery:(Epos2DeviceInfo *)deviceInfo {
    NSLog(@"onDiscovery: %@", [deviceInfo getIpAddress]);
    NSDictionary *info = @{@"ipAddress" : [deviceInfo getIpAddress],
                           @"deviceType": [NSNumber numberWithInt:[deviceInfo getDeviceType]],
                           @"target": [deviceInfo getTarget],
                           @"deviceName": [deviceInfo getDeviceName],
                           @"macAddress": [deviceInfo getMacAddress],
                           @"bdAddress": [deviceInfo getBdAddress],
                           };
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:info];
    [result setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:result callbackId:self.discoverCallbackId];
}

- (void)stopDiscover:(CDVInvokedUrlCommand *)command {
    NSLog(@"stopDiscover()");
    int result = EPOS2_SUCCESS;

    while (YES) {
        result = [Epos2Discovery stop];

        if (result != EPOS2_ERR_PROCESSING) {
            break;
        }
    }

    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in stopDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in stop discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

-(void)connectPrinter:(CDVInvokedUrlCommand *)command {
    initWithPrinterSeries:printerSeries =   [[command.arguments objectAtIndex:0] intValue];
    NSString* target = [command.arguments objectAtIndex:1];
    
    NSLog(@"connectPrinter(), ip=%@", target);

    printer = [[Epos2Printer alloc] initWithPrinterSeries:printerSeries lang:lang];

    [printer setReceiveEventDelegate:self];

    if (printer == nil) {
        NSLog(@"error in connectPrinter(), printer object not found");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer. Printer object not found"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    int result = [printer connect:[NSString stringWithFormat:@"TCP:%@", target] timeout:EPOS2_PARAM_DEFAULT];
    
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in connectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    result = [printer beginTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in beginTransaction()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in begin transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        [printer disconnect];
        return;
    }

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)disconnectPrinter:(CDVInvokedUrlCommand *)command {

    int result = [printer endTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in end transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }

    result = [printer disconnect];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in disconnect printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
    
    [self finalizeObject];

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId {
    NSLog(@"onPtrReceive");
    NSLog(@"code: %d", code);
    NSLog(@"status: %@", status);
    NSLog(@"printJobId: %@", printJobId);

    [self disconnectPrinter:nil];
}

- (void)addText:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    NSString* printData = [command.arguments objectAtIndex:0];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addText:printData]];
    }];
}

- (void)addTextAlign:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int data = [[command.arguments objectAtIndex:0] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addTextAlign:data]];
    }];
}

- (void)addTextSize:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int width = [[command.arguments objectAtIndex:0] intValue];
    int height = [[command.arguments objectAtIndex:1] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addTextSize: width height:height]];
    }];
}

- (void)addFeedLine:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int line = [[command.arguments objectAtIndex:0] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addFeedLine:line]];
    }];
}

- (void)addBarcode:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    NSString* data = [command.arguments objectAtIndex:0];
    int type = [[command.arguments objectAtIndex:1] intValue];
    int hri = [[command.arguments objectAtIndex:2] intValue];
    int font = [[command.arguments objectAtIndex:3] intValue];
    int width = [[command.arguments objectAtIndex:4] intValue];
    int height = [[command.arguments objectAtIndex:5] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addBarcode:data type:type hri:hri font:font width:width height:height]];
    }];
}

- (void)addCut:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int type = [[command.arguments objectAtIndex:0] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addCut:0]];
    }];
}

- (void)addTextStyle:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int reverse = [[command.arguments objectAtIndex:0] intValue];
    int ul = [[command.arguments objectAtIndex:1] intValue];
    int em = [[command.arguments objectAtIndex:2] intValue];
    int color = [[command.arguments objectAtIndex:3] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addTextStyle:reverse ul:ul em:em color:color]];
    }];
}

- (void)addPulse:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    int drawer = [[command.arguments objectAtIndex:0] intValue];
    int time = [[command.arguments objectAtIndex:1] intValue];
    
    [self.commandDelegate runInBackground:^{
        [self processResult: [printer addPulse:drawer time:time]];
    }];
}

- (void)print:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;

    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [self.commandDelegate runInBackground:^{
        [self processResult: [printer sendData:EPOS2_PARAM_DEFAULT]];
    }];
}

- (void)processResult: (int) result {
    if (result != EPOS2_SUCCESS) {
        [printer disconnect];
        
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in print data."];
        [self.commandDelegate sendPluginResult:result callbackId:self.printCallbackId];
        return;
    }
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.printCallbackId];
    return;
}

- (void)finalizeObject
{
    if (printer == nil) {
        return;
    }
    
    [printer clearCommandBuffer];
    
    [printer setReceiveEventDelegate:nil];
    
    printer = nil;
}

@end
